backend/controller/dal/lease.go

```diff
@@ -75,28 +80,53 @@ func (l *Lease) Release() error {
 // AcquireLease acquires a lease for the given key.
 //
 // Will return ErrConflict if the lease is already held by another controller.
-func (d *DAL) AcquireLease(ctx context.Context, key leases.Key, ttl time.Duration) (leases.Lease, error) {
+func (d *DAL) AcquireLease(ctx context.Context, key leases.Key, ttl time.Duration, metadata optional.Option[any]) (leases.Lease, context.Context, error) {
 	if ttl < time.Second*5 {
-		return nil, fmt.Errorf("lease TTL must be at least 5 seconds")
+		return nil, nil, fmt.Errorf("lease TTL must be at least 5 seconds")
 	}
-	idempotencyKey, err := d.db.NewLease(ctx, key, ttl)
+	var metadataBytes []byte
+	if md, ok := metadata.Get(); ok {
+		var err error
+		metadataBytes, err = json.Marshal(md)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to marshal lease metadata: %w", err)
+		}
+	}
+	idempotencyKey, err := d.db.NewLease(ctx, key, ttl, metadataBytes)
 	if err != nil {
-		return nil, translatePGError(err)
+		return nil, nil, translatePGError(err)
 	}
-	return d.newLease(ctx, key, idempotencyKey, ttl), nil
+	leaseCtx, lease := d.newLease(ctx, key, idempotencyKey, ttl)
+	return leaseCtx, lease, nil
 }
 
-func (d *DAL) newLease(ctx context.Context, key leases.Key, idempotencyKey uuid.UUID, ttl time.Duration) *Lease {
+func (d *DAL) newLease(ctx context.Context, key leases.Key, idempotencyKey uuid.UUID, ttl time.Duration) (*Lease, context.Context) {
+	leaseCtx, cancelCtx := context.WithCancel(ctx)
```

Comment:
Don't have a separate `leaseCtx`, just use `ctx` pass that through to `lease.renew()`, and don't store it. Storing contexts in structs is (in general) a code smell.

Reply:
I've changed it to be set to `ctx` and passed it into `renew`, but we aren't storing the lease ctx here.

I am storing a context in [Coordinator](https://github.com/TBD54566975/ftl/blob/matt2e%2Fleader/backend/controller/leader/coordinator.go#L40) though. Is there a better way to do it there?
Originally I had `Get()` pass in a context but the ctx provided in some cases ended up cancelling before when you'd want the leader or follower to end.

Reply:
Sometimes it's not possible to avoid, I've found.